components/brave_today/browser/direct_feed_controller.cc

```diff
@@ -97,11 +100,62 @@ void ParseFeedDataOffMainThread(const GURL& feed_url,
 }  // namespace
 
 DirectFeedController::DirectFeedController(
+    PrefService* prefs,
     scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
-    : url_loader_factory_(url_loader_factory) {}
+    : prefs_(prefs), url_loader_factory_(url_loader_factory) {}
 
 DirectFeedController::~DirectFeedController() = default;
 
+bool DirectFeedController::AddDirectFeedPref(
+    const GURL& feed_url,
+    const std::string& title,
+    const absl::optional<std::string>& id) {
+  // Check if feed url already exists
+  auto* existing_feeds = prefs_->GetDictionary(prefs::kBraveTodayDirectFeeds);
+  for (const auto kv : existing_feeds->DictItems()) {
+    // Non dict values will be flagged as an issue elsewhere.
+    if (!kv.second.is_dict())
+      continue;
+
+    auto existing_url =
+        *kv.second.FindStringKey(prefs::kBraveTodayDirectFeedsKeySource);
+    if (GURL(existing_url) == feed_url.spec()) {
```

Comment:
Should there be any sanitization here? (ex: stripping query string / fragments). Ideally, sites shouldn't be adding those and I can't remember if `GURL`'s `.spec` excludes those

Reply:
I don't know about in general, but I don't think we should do anything to the URL before comparing it. Both of these URLs will be something that's come back from Brave News, and I think by stripping query params we'd be risking the URLs not lining up.

Maybe in future we could try stripping query params (they're a potential tracking vector) but I'd be quite worried about breaking URLs. In addition, if they really want to track users, they could add a unique path to the url, instead of a GUID, or serve from a unique subdomain.